pkg/lang/csharp/plugin_expose_test.go

```diff
@@ -0,0 +1,808 @@
+			units: map[string][]srcFile{
+				"main": {
+					{
+						Path:    "Startup.cs",
+						Content: controllerMappingStartupClass,
+					},
+					{
+						Path: "MyController.cs",
+						Content: `
+						using Microsoft.AspNetCore.Mvc;
+						[Route("/api/[controller]")]
+						public class MyController {
+							[Route("child")]
+							[Route("/root/child")]
+							[AcceptVerbs("GET")]
+							[AcceptVerbs("DELETE", Route="/del")]
+							[HttpGet]
+							[HttpDelete,HttpPut]
+							public void action() {}
+						}
+						`,
+					},
+				},
+			},
+			expectedGateways: []gateway{
+				{
+					Name: "my-gateway",
+					Routes: []routeMethodPath{
+						{Verb: core.VerbGet, Path: "/api/my"},
+						{Verb: core.VerbDelete, Path: "/api/my"},
+						{Verb: core.VerbPut, Path: "/api/my"},
+						{Verb: core.VerbGet, Path: "/api/my/child"},
+						{Verb: core.VerbGet, Path: "/root/child"},
+						{Verb: core.VerbDelete, Path: "/del"},
+					},
+				},
+			},
+			expectedDeps: []string{
+				"my-gateway:main",
+			},
+		},
+		{
+			name: "Optional last path params result in required and optional routes",
+			units: map[string][]srcFile{
+				"main": {
+					{
+						Path:    "Startup.cs",
+						Content: controllerMappingStartupClass,
+					},
+					{
+						Path: "MyController.cs",
+						Content: `
+						using Microsoft.AspNetCore.Mvc;
+						public class MyController {
+							[Route("/required/{optional?}")]
+							[AcceptVerbs("GET")]
+							public void action() {}
+						}
+						`,
+					},
+				},
+			},
+			expectedGateways: []gateway{
+				{
+					Name: "my-gateway",
+					Routes: []routeMethodPath{
+						{Verb: core.VerbGet, Path: "/required"},
+						{Verb: core.VerbGet, Path: "/required/:optional"},
+					},
+				},
+			},
+			expectedDeps: []string{
+				"my-gateway:main",
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert := assert.New(t)
+			result := &core.CompilationResult{}
+			for uName, files := range tt.units {
+				unit := &core.ExecutionUnit{
+					Name:       uName,
+					Executable: core.NewExecutable(),
+				}
+				for _, f := range files {
+					sf, err := core.NewSourceFile(f.Path, strings.NewReader(f.Content), Language)
+					if !assert.NoError(err) {
+						return
+					}
+					unit.AddSourceFile(sf)
+				}
+				result.Add(unit)
+			}
+			deps := &core.Dependencies{}
+			expose := Expose{}
+			err := expose.Transform(result, deps)
+			if !assert.NoError(err) {
+				return
+			}
+
+			gateways := core.GetResourcesOfType[*core.Gateway](result)
+			assert.Equal(len(tt.expectedGateways), len(gateways))
+
+			sort.Slice(gateways, func(i, j int) bool {
+				return gateways[i].Name < gateways[j].Name
+			})
+			sort.Slice(tt.expectedGateways, func(i, j int) bool {
+				return tt.expectedGateways[i].Name < tt.expectedGateways[j].Name
+			})
+
+			for _, gw := range gateways {
+				sort.Slice(gw.Routes, func(i, j int) bool {
```

Comment:
We should consider moving this into the compiler if the Routes (or anything else in list format) is in an inconsistent order.

Reply:
The purpose of sorting here is to eliminate the need to figure out the correct order for expectations.